pkg/serviceaccount/claims.go

```diff
@@ -76,3 +79,102 @@ func Claims(sa core.ServiceAccount, pod *core.Pod, secret *core.Secret, expirati
 	}
 	return sc, pc
 }
+
+func NewValidator(audiences []string, lookup bool, getter ServiceAccountTokenGetter) Validator {
+	return &validator{
+		auds:   audiences,
+		lookup: lookup,
+		getter: getter,
+	}
+}
+
+type validator struct {
+	auds   []string
+	lookup bool
+	getter ServiceAccountTokenGetter
+}
+
+var _ = Validator(&validator{})
+
+func (v *validator) Validate(_ string, public *jwt.Claims, privateObj interface{}) (string, string, string, error) {
+	private, ok := privateObj.(*privateClaims)
+	if !ok {
+		glog.Errorf("jwt validator expected private claim of type *privateClaims but got: %T", privateObj)
+		return "", "", "", errors.New("Token could not be validated.")
+	}
+	err := public.Validate(jwt.Expected{
+		Audience: v.auds,
+		Time:     now(),
+	})
+	switch {
+	case err == nil:
+	case err == jwt.ErrExpired:
+		return "", "", "", errors.New("Token has expired.")
+	case err == jwt.ErrInvalidAudience:
+		return "", "", "", errors.New("Token is invalid for this audience.")
+	default:
+		glog.Errorf("unexpected validation error: %T", err)
+		return "", "", "", errors.New("Token could not be validated.")
+	}
+
+	if v.lookup {
+		namespace := private.Kubernetes.Namespace
+		saref := private.Kubernetes.Svcacct
+		podref := private.Kubernetes.Pod
+		secref := private.Kubernetes.Secret
+		// Make sure service account still exists (name and UID)
+		serviceAccount, err := v.getter.GetServiceAccount(namespace, saref.Name)
```

Comment:
I think this chunk would be more readable w/ function calls for the validation of each object. E.g.:
```
if err != v.checkServiceAccountRef(namespace, saref.Name); err != nil {
	return "", "", "", err
}
if secref != nil {
	if err != v.checkSecretRef(namespace, secref.Name); err != nil {
		return "", "", "", err
	}
}
...
```

Then it's more obvious to see the common structure, and the fact that saref is always important, but the others may not be there.